### PR TITLE
Fix MSVC VST compilation

### LINKED ISSUE #4103: 64 bit VST on Linux/Wine
Wine supports 64-bit since wine-1.2 but it doesn't seem to be working under linux/wine/lmms.
https://github.com/LMMS/lmms/issues/4102, https://lmms.io/forum/viewtopic.php?f=7&t=26647#p51235

- [ ] Assess the status/quality of 64-bit VST on Linux/Wine
- [ ] If possible, activate 64-bit VST on Linux/Wine
- [ ] Warn if needed/possible when running 64-bit VST on Linux/Wine
- [ ] Update documentation

---

Wine 3.0 is in the works and [rc5 was recently released](https://www.winehq.org/news/2018010501).
https://www.winehq.org/

### PR DESCRIPTION
**Pull request by lukas-w**
This PR fixes VST compilation with MSVC (see #3999). Changes made include:

1. A lot of changes to CMake scripts had to be made. `RemoteVstPlugin` is now built using `ExternalProject_Add` when the target platform doesn't match, and is named `RemoteVstPlugin32` and `RemoteVstPlugin64` for 32bit and 64bit respectively. WineGCC isn't used directly any more, but is passed to the CMake external project with `-DCMAKE_CXX_COMPILER=...`-
2. Use [`GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) instead of the hand-written `export.h`. The `EXPORT` define was renamed to `LMMS_EXPORT`.
3. Support for [vcpkg ](https://github.com/Microsoft/vcpkg) and numerous other fixes were added by @justnope in #4352. Thank you and please come back! :heart: 
4. I cherry-picked @Reflexe's SDL2 fixes from #3947 to make LMMS usable. Thanks!
5. `RemoteVstPlugin.cpp` was ported to use C++11 threads instead of POSIX threads. [mingw-std-threads](https://github.com/meganz/mingw-std-threads) is included as a submodule and used for MinGW versions that don't support C++11 threads natively.
6. Vestige reads DLL's file headers to determine whether to use `RemoteVstPlugin32` or `RemoteVstPlugin64`. Thanks @tresf for the [suggestion](https://github.com/LMMS/lmms/issues/4103#issuecomment-357972162).

As a positive side-effect, this enables 64bit VSTs for Linux (fixes #4103). Here's an AppImage: https://transfer.sh/LFwDB/lmms-1.2.0-rc5.137-linux-x86_64.AppImage

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -100,7 +100,7 @@ jobs:
           command: |
             cd build
             make install
-            make appimage
+            make appimage || cat appimage.log
```
Comment (tresf) [inline .circleci/config.yml]: 👍

File: cmake/linux/winegcc_wrapper.in [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
+# Wrapper script for winegcc to remove .exe file ending automatically
+# appended by winebuild.
+# Usage: winegcc <args ...>
```
Comment (tresf) [inline cmake/linux/winegcc_wrapper.in]: Your usage says `winegcc`.  I'm sure this was intentional, but more accurately...

```bash
# Usage: winegcc_wrapper <args ...>

# --- OR ---
# Usage:
#    CONFIGURE_FILE("path/to/winegcc_wrapper.in winegcc_wrapper @ONLY)
#    SET(WINEGCC "path/to/winegcc_wrapper")
#    winegcc <args ...>
```

Probably nitpicking, but didn't know if it was a typo or intentional. :)

**Comment by lukas-w** [comment]
I'll be merging this shortly if there are no objections. This changes a lot of things in CMake, so I it's possible it breaks the build for some Linux users. Feel free to complain to me if that's the case.

File: plugins/vst_base/VstPlugin.cpp [hunk 3]
```diff
@@ -91,20 +147,32 @@ VstPlugin::VstPlugin( const QString & _plugin ) :
...
+		qCritical() << "Error while determining PE file's machine type: " << e.what();
+		machineType = PE::MachineType::unknown;
+	}
+
+	switch(machineType)
 	{
-		m_badDllFormat = false;
-		tryLoad( "32/RemoteVstPlugin32" );
+	case PE::MachineType::amd64:
+		tryLoad( "RemoteVstPlugin64" );
```
Comment (PhysSong) [inline plugins/vst_base/VstPlugin.cpp]: If I understand correctly, this will try to load 64bit DLL in 32bit system using nonexistent `RemoteVstPlugin64`. It will print an error message, but Windows users can't see that.
Do we need to try this on 32bit LMMS?

Comment (lukas-w) [reply plugins/vst_base/VstPlugin.cpp]: This is correct, there's no visible error message in the GUI when loading fails. This PR doesn't change that.

**Comment by lukas-w** [comment]
Anything left to change @PhysSong or should I merge?

File: cmake/modules/FindWine.cmake [hunk 4]
```diff
@@ -24,6 +31,10 @@ STRING(REPLACE " " ";" WINEBUILD_FLAGS "${WINEBUILD_OUTPUT}")
 
 FOREACH(FLAG ${WINEBUILD_FLAGS})
 	IF("${FLAG}" MATCHES "libwinecrt0.a.*")
+		STRING(REGEX REPLACE "/wine/libwinecrt0.a.*" "" FLAG "${FLAG}")
+
+		SET(WINE_64_LIBRARY_DIR "${FLAG}/")
```
Comment (PhysSong) [inline cmake/modules/FindWine.cmake]: Unusual, but it is possible to use 32-bit WINE toolchain `wine32-tools` on 64-bit Debian/Ubuntu. In that case, 64-bit VST compilation needs the inverse replacing(32-bit toolchain uses 32-bit libraries, even if `-m64` is specified).

File: plugins/CMakeLists.txt [hunk 5]
```diff
@@ -75,8 +76,8 @@ IF("${PLUGIN_LIST}" STREQUAL "")
 		stereo_enhancer
 		stereo_matrix
 		stk
-		vestige
 		vst_base
+		vestige
```
Comment (PhysSong) [inline plugins/CMakeLists.txt]: I think there's no reason to move this.

File: include/stdshims.h [hunk 6]
```diff
@@ -7,8 +7,10 @@
 #include <memory>
 #include <utility>
 
-#if (__cplusplus >= 201402L)
+#if (__cplusplus >= 201402L || _MSC_VER)
```
Comment (PhysSong) [inline include/stdshims.h]: It looks VS 2012 incompatible.
https://github.com/LMMS/lmms/pull/4352#discussion_r191983580

File: plugins/vst_base/RemoteVstPlugin.cpp [hunk 7]
```diff
@@ -767,15 +783,12 @@ bool RemoteVstPlugin::load( const std::string & _plugin_file )
 {
 	if( ( m_libInst = LoadLibrary( _plugin_file.c_str() ) ) == NULL )
 	{
-		// give VstPlugin class a chance to start 32 bit version of RemoteVstPlugin
-		if( GetLastError() == ERROR_BAD_EXE_FORMAT )
-		{
-			sendMessage( IdVstBadDllFormat );
-		}
+		DWORD error	= GetLastError();
```
Comment (PhysSong) [inline plugins/vst_base/RemoteVstPlugin.cpp]: Minor, but a tab before `=` doesn't look good here.

File: plugins/vst_base/VstPlugin.cpp (hunk 3 above)
Comment (PhysSong) [reply plugins/vst_base/VstPlugin.cpp]: I think error message issue can be fixed later, but do we ever need to try to load VST with `RemoteVstPlugin64` on 32bit systems?

File: src/core/RemotePlugin.cpp [hunk 8]
```diff
@@ -62,7 +63,10 @@ void ProcessWatcher::run()
 	{
 		fprintf( stderr,
 				"remote plugin died! invalidating now.\n" );
+
+#ifndef SYNC_WITH_SHM_FIFO
```
Comment (PhysSong) [inline src/core/RemotePlugin.cpp]: Could you explain why did you add this check?

File: cmake/modules/FindWine.cmake (hunk 4 above)
Comment (lukas-w) [reply cmake/modules/FindWine.cmake]: Is that something we have supported before and that this PR breaks?

File: plugins/CMakeLists.txt (hunk 5 above)
Comment (lukas-w) [reply plugins/CMakeLists.txt]: There is, `vestige` links to `vstbase`, so `vst_base` needs to be processed first. Otherwise, the target doesn't exist when `vestige` tries to link to it.

File: plugins/vst_base/VstPlugin.cpp (hunk 3 above)
Comment (lukas-w) [reply plugins/vst_base/VstPlugin.cpp]: Theoretically, we could ship a `RemoteVstPlugin64` with our 32bit Windows installers for users who install 32bit LMMS because they're unaware of the difference. In that case, it's good to try `RemoteVstPlugin64`. In any other case, there's no harm done here. `tryLoad` will discover that `RemoteVstPlugin64` doesn't exist and will abort, at which point we can easily add an error message later.

File: src/core/RemotePlugin.cpp (hunk 8 above)
Comment (lukas-w) [reply src/core/RemotePlugin.cpp]: If you look at the commit that added this change (https://github.com/LMMS/lmms/pull/4421/commits/9db8cbfb31f9b6d3e754887a6280770e3d3e2bcd#diff-b6e06d3d61c47ede770ac49ad9ea8f9e), you can see that I removed an `invalidate()` call guarded by the same check in `processFinished`. I removed it because it prevented receiving messages after the plugin died (related: https://github.com/LMMS/lmms/issues/4103#issuecomment-357974063, #4371, 4fd8ecd7e4084aedbd39900ffe5cfd107ceba1a0). I figured `@jasp00` had a reason to add the check though and simply missed this `invalidate` call, because I couldn't see a point in doing the `invalidate` twice. So I added the checks here because I removed them together with the `invalidate` call in `processFinished`. I'm not sure about this though, it's mostly guesswork. Maybe @DomClark or @Wallacoloo have any insight on this?

File: plugins/vst_base/VstPlugin.cpp (hunk 3 above)
Comment (PhysSong) [reply plugins/vst_base/VstPlugin.cpp]: Got it, I didn't think about that.

File: src/core/RemotePlugin.cpp (hunk 8 above)
Comment (PhysSong) [reply src/core/RemotePlugin.cpp]: Since `SYNC_WITH_SHM_FIFO` is defined on Windows, this check will prevent LMMS from invalidating plugin. There's no other calls to `invalidate` on Windows, so it may cause crashes/hangs when remote plugin crashes.
Also, dropping the call in `processFinished` may result in delaying invalidation because it checks process state every 200ms. I guess call was added for immediate invalidation.
I guess we can somehow work around the delay and/or the IPC issue.

Comment (lukas-w) [reply src/core/RemotePlugin.cpp]: Should I remove the check? Re-adding the call in `processFinished` will render @Justnope's/@Wallacoloo's 4fd8ecd7e4084aedbd39900ffe5cfd107ceba1a0 useless again.

Comment (PhysSong) [reply src/core/RemotePlugin.cpp]: > Should I remove the check?

I think yes. There's no reason to check that here.
> Re-adding the call in `processFinished` (...) 4fd8ecd useless

It does, but can't we work around this? Though it's not necessary for now and might be hard, I don't think it's impossible. I don't want to block merging this because of the call though.

File: plugins/vst_base/VstPlugin.cpp (hunk 3 above)
Comment (PhysSong) [reply plugins/vst_base/VstPlugin.cpp]: I think we may try to detect processor in runtime if possible, though it looks like out of scope.
Qt provides [`QSysInfo::currentCpuArchitecture`](http://doc.qt.io/qt-5/qsysinfo.html#currentCpuArchitecture) as of Qt 5.4, and it's also possible to use OS-dependent detection.

Comment (lukas-w) [reply plugins/vst_base/VstPlugin.cpp]: I opened #4447 to track it.

File: src/core/RemotePlugin.cpp (hunk 8 above)
Comment (lukas-w) [reply src/core/RemotePlugin.cpp]: > It does, but can't we work around this?

We probably can, but I have no idea how to do it.

Anyway, I'll revert both changes because this PR doesn't depend on them any more since the `VstBadDllFormat` message was replaced by the PE file detection.

**Comment by lukas-w** [comment]
@PhysSong I fixed the white-space error and removed the `invalidate`-related changes so they can be discussed in #4371.

Thank you for reviewing! I'll merge once you approve.

File: plugins/vst_base/RemoteVstPlugin.cpp [hunk 9]
```diff
@@ -675,11 +691,11 @@ void RemoteVstPlugin::init( const std::string & _plugin_file )
 
 
 
-static void close_check( int fd )
+static void close_check( FILE* fd )
```
Comment (PhysSong) [inline plugins/vst_base/RemoteVstPlugin.cpp]: Minor, but it's not a file descriptor anymore. Might `fd` be confusing?

File: .circleci/config.yml (hunk 1 above)
Comment (PhysSong) [reply .circleci/config.yml]: Won't this mark the build as successful if AppImage creation is failed? I think it's fine if desired though.

**Comment by PhysSong** [comment]
@lukas-w For now, I can't find any more issues. Once you address my comments, I'd say let's merge this.
Anyway, I wonder if other Devs such as @DomClark want to review this one.

File: .circleci/config.yml (hunk 1 above)
Comment (tresf) [reply .circleci/config.yml]: Perhaps `make appimage || cat appimage.log && false`

Comment (PhysSong) [reply .circleci/config.yml]: > Perhaps `make appimage || cat appimage.log && false`

I think there should be a pair of parenthesis:  `make appimage || (cat appimage.log && false)`
See also: https://unix.stackexchange.com/questions/88850/

Comment (lukas-w) [reply .circleci/config.yml]: I think if appimage creation fails, it will fail one line later in `cp ./lmms-*.AppImage /tmp/artifacts/`. I'll still add the `&& false` to make sure.

Comment (PhysSong) [reply .circleci/config.yml]: > I think if appimage creation fails, it will fail one line later in `cp ./lmms-*.AppImage /tmp/artifacts/`.

True, I just missed that.

**Comment by lukas-w** [comment]
@PhysSong Fixed the potential AppImage problem and the confusing `FILE*` variable names. Note that the `FILE*` changes in this PR are in conflict with #4401. Feel free to merge your PR first, I can take care of the merge conflicts then.

**Comment by lukas-w** [comment]
Rebased to squash some commits and because I failed to attribute ad4c4f0273d827fac8e5c47d18145da020250cd7 to @Reflexe and merged via 846a2afeaa9bc20f160a7cf086c989273505c890.

File: plugins/vst_base/VstPlugin.cpp [hunk 10]
```diff
@@ -66,27 +69,93 @@
...
+	i386 = 0x14c,
+};
+
+class FileInfo
+{
+public:
+	FileInfo(QString filePath)
+		: m_file(filePath)
+	{
+		m_file.open(QFile::ReadOnly);
```
Comment (PhysSong) [inline plugins/vst_base/VstPlugin.cpp]: This file open can fail if the path passed to `VstPlugin` is relative. So I can't load some VSTs...
There should be a path check logic like one in `VstPlugin::tryLoad`.

File: cmake/modules/FindWine.cmake [hunk 11]
```diff
@@ -33,12 +44,19 @@ FOREACH(FLAG ${WINEBUILD_FLAGS})
 		# WineHQ (/opt/wine-stable, /opt/wine-devel, /opt/wine-staging)
 		STRING(REPLACE "/lib64/wine/" "/lib/wine/" FLAG "${FLAG}")
```
Comment (PhysSong) [inline cmake/modules/FindWine.cmake]: This logic doesn't work anymore because we drop `/wine/` before it and causes #4501.
Some possible solutions:

- Use a regular expression to match trailing `lib64`
- Remove `/wine/libwinecrt0.a.*` after replacing
- Always replace `lib64` with `lib` if it doesn't break other platforms

File: plugins/vst_base/VstPlugin.cpp (hunk 10 above)
Comment (lukas-w) [reply plugins/vst_base/VstPlugin.cpp]: Fixed via 7ddca851d1b18bbc64b7d1ee15078b0daa8501c9

File: cmake/modules/FindWine.cmake (hunk 11 above)
Comment (lukas-w) [reply cmake/modules/FindWine.cmake]: Thanks for tracking this down. Fixed with 515fefa6ce734fcc5cf119b8eff8c3c0db65cdfc.
